core/src/main/scala/quasar/destination/gbq/GBQDestination.scala

```diff
@@ -0,0 +1,213 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.destination.gbq
+
+import slamdata.Predef._
+
+import quasar.api.destination.{Destination, DestinationError, DestinationType, ResultSink}
+import quasar.api.destination.DestinationError.InitializationError
+import quasar.connector.{MonadResourceErr, ResourceError}
+import quasar.api.push.RenderConfig
+import quasar.api.resource.ResourceName
+import quasar.api.table.{ColumnType, TableColumn}
+
+import argonaut._, Argonaut._
+
+import cats.ApplicativeError
+import cats.effect.{Concurrent, ContextShift, Resource, Sync}
+import cats.implicits._
+
+import com.google.auth.oauth2.AccessToken
+
+import eu.timepit.refined.auto._
+
+import fs2.Stream
+
+import org.http4s.{
+  AuthScheme,
+  Credentials,
+  EntityEncoder,
+  MediaType,
+  Method,
+  Request,
+  Status,
+  Uri}
+import org.http4s.argonaut.jsonEncoderOf
+import org.http4s.client.Client
+import org.http4s.headers.{Authorization, `Content-Type`, Location}
+
+import org.slf4s.Logging
+
+import scalaz.NonEmptyList
+
+import shims._
+
+final class GBQDestination[F[_]: Concurrent: ContextShift: MonadResourceErr](
+    client: Client[F], 
+    config: GBQConfig,
+    sanitizedConfig: Json) extends Destination[F] with Logging {
+
+  def destinationType: DestinationType = DestinationType("gbq", 1L)
+
+  def sinks: NonEmptyList[ResultSink[F]] = NonEmptyList(gbqSink)
+
+  private def gbqSink: ResultSink[F] = 
+    ResultSink.csv[F](RenderConfig.Csv()) { (path, columns, bytes) =>
+      val gbqSchema = tblColumnToGBQSchema(columns)
+      
+      val tableNameF = path.uncons match {
+        case Some((ResourceName(name), _)) => name.pure[F]
+        case _ => MonadResourceErr[F].raiseError[String](
+          ResourceError.notAResource(path))
+      }
+
+      for {
+        tableName <- Stream.eval[F, String](tableNameF)
+        gbqJobConfig = formGBQJobConfig(gbqSchema, config, tableName)
+        accessToken <- Stream.eval(GBQAccessToken.token(config.authCfg.getBytes("UTF-8")))
+        _ <- Stream.eval((mkDataset(client, accessToken, config)))
+        eitherloc <- Stream.eval(mkGbqJob(client, accessToken, gbqJobConfig))
+        _ <- Stream.eval(
+          Sync[F].delay(
+            log.info(s"(re)creating ${config.project}.${config.datasetId}.${tableName} with schema ${columns.show}")))
+        _ <- eitherloc match {
+            case Right(locationUri) => {
+              upload(client, bytes, locationUri, accessToken)
+            }
+            case Left(e) =>
+              Stream.eval(Sync[F].delay(ApplicativeError[F, Throwable].raiseError(
+                new RuntimeException(s"No Location URL from returned from job: $e"))))
```

Comment:
The above doesn't actually raise an error, because you're creating a `raiseError` effect within `delay` (which is what you use for *side-effects*, which this is not) and then throwing it away. Also it results in value discard, which probably means that the compiler warnings are being ignored?

Reply:
```
scala> Stream.eval(Sync[IO].delay(ApplicativeError[IO, Throwable].raiseError(new RuntimeException("aaaa")))).compile.toList.unsafeRunSync()

res12: List[cats.effect.IO[Nothing]] = List(IO(throw java.lang.RuntimeException: aaaa))
```
versus
```
scala> Stream.eval(ApplicativeError[IO, Throwable].raiseError(new RuntimeException("aaaa"))).compile.toList.unsafeRunSync()

java.lang.RuntimeException: aaaa
  ... 36 elided
```

`ApplicativeError#raiseError` returns `F[A]`. The original code is wrapping that `F[A]` again in`F` via `Sync[F].delay`. This outer `F` is then the one that is evaluated in `Stream.eval`. But the inner `F` is the one that incorporates the `Throwable`, and so the exception isn't being raised. Thus omitting the `Sync[F].delay` does the trick.

Reply:
Oh, @cryogenian just commented suggesting `Stream.raiseError`. That is more succinct than the suggestion I had above, so I'd use that 👍